### ISSUE TITLE
How do I access per-batch training data?

### ISSUE THREAD

**Issue by LarryThermo** [GRAAL-Research/poutyne#74]
First of all I have used Poutyne on several projects very successfully and am appreciative of the ease of use and how much it simplifies my Pytorch training code.

I wish to incorporate adversarial loss during my training, along the lines of the Usage example in https://github.com/lyakaap/VAT-pytorch . In order to compute adversarial loss I need access to the model and per-batch training data before the forward pass. How would one do this in Poutyne or in Poutyne callback? Thanks, Lars

**Comment by freud14** [comment]
Hi,
Your network can return multiple values as a tuple. So I suggest you to return the VATLoss in addition to the regular output of the network. Then, you can define your loss function inside your network has follow:
```python
class MyNetwork(nn.Module):
    ....
    def forward(self, ....):
        ....
        return output, lds

    def loss_function(self, output, target):
        output, lds = output
        return F.cross_entropy(output, target) + self.alpha * lds

net = MyNetwork(...)
model = Model(net, loss_function=net.loss_function, ....)
```
Let me know whether that does the trick.

Thank you.

**Comment by LarryThermo** [comment]
Hi freud14,

Thanks for responding.

I am confused about the exact implementation of MyNetwork.Forward and how to compute the lds inside it.

This is because lds in the VAT-pytorch example comes from calling VATLoss.vat_loss.__call__ which calls  VATLoss.forward function that in turn calls model.forward. I put a "HERE" comment where model.__call__  indirectly calls model.forward (below). I conclude this would lead to an infinite loop.

My thinking at this point was that maybe Callback.on_batch_begin could be enhanced to include "x" in the logs dictionary. But thank you for any input you can give.

Lars



# VATLoss implementation reproduced from the vat.py file for clarity of discussion:
class VATLoss(nn.Module):

    def __init__(self, xi=10.0, eps=1.0, ip=1):
        """VAT loss
        :param xi: hyperparameter of VAT (default: 10.0)
        :param eps: hyperparameter of VAT (default: 1.0)
        :param ip: iteration times of computing adv noise (default: 1)
        """
        super(VATLoss, self).__init__()
        self.xi = xi
        self.eps = eps
        self.ip = ip

    def forward(self, model, x):
        with torch.no_grad():
            pred = F.softmax(model(x), dim=1) # HERE is where model.forward is called indirectly

        # prepare random unit tensor
        d = torch.rand(x.shape).sub(0.5).to(x.device)
        d = _l2_normalize(d)

        with _disable_tracking_bn_stats(model):
            # calc adversarial direction
            for _ in range(self.ip):
                d.requires_grad_()
                pred_hat = model(x + self.xi * d)
                logp_hat = F.log_softmax(pred_hat, dim=1)
                adv_distance = F.kl_div(logp_hat, pred, reduction='batchmean')
                adv_distance.backward()
                d = _l2_normalize(d.grad)
                model.zero_grad()
    
            # calc LDS
            r_adv = d * self.eps
            pred_hat = model(x + r_adv)
            logp_hat = F.log_softmax(pred_hat, dim=1)
            lds = F.kl_div(logp_hat, pred, reduction='batchmean')

        return lds

**Comment by LarryThermo** [comment]
My bad for closing this issue prematurely and not understanding the github discussion system.

**Comment by freud14** [comment]
Hi, 

>This is because lds in the VAT-pytorch example comes from calling VATLoss.vat_loss.call which calls VATLoss.forward function that in turn calls model.forward. I put a "HERE" comment where model.call indirectly calls model.forward (below). I conclude this would lead to an infinite loop.

Indeed, you cannot do this as you describe. What is a module wrapping your network. Here is an example of code (I didn't test it, syntax might not be correct) that does it:

```python
class NetworkWrapper(nn.Module):
    def __init__(self, my_network):
        super().__init__()
        self.my_network = my_network
        self.vat_loss = VATLoss()

    def forward(self, x):
        output = self.my_network(x)
        lds = self.vat_loss(self.my_network, x)
        return output, lds

    def loss_function(self, output, target):
        output, lds = output
        return F.cross_entropy(output, target) + self.alpha * lds

net = MyNetwork(...)
wrapper = NetworkWrapper(net)
model = Model(wrapper, loss_function=wrapper.loss_function, ....)
```

> My thinking at this point was that maybe Callback.on_batch_begin could be enhanced to include "x" in the logs dictionary. But thank you for any input you can give.

You cannot do that since you want your VATLoss to be counted in your gradient. Also, you're not supposed to modify the log dictionnary. If you want to have a metric for your VATLoss, you can do something like that:

```python
class NetworkWrapper(nn.Module):
    ...

    def vatloss(self, output, target):
        output, lds = output
        return self.alpha * lds

net = MyNetwork(...)
wrapper = NetworkWrapper(net)
model = Model(wrapper, 
              loss_function=wrapper.loss_function, 
              batch_metrics=[wrapper.vatloss], ....)
```

I hope my reply is clear.

Thank you.

**Comment by larry-rystrom** [comment]
Thank you for the suggestion.

**Comment by freud14** [comment]
Hi,
Did it solve your problem? If so, you can close the issue.
Thank you.
